lls/cache.h

```diff
@@ -0,0 +1,95 @@
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _GATEKEEPER_LLS_CACHE_H_
+#define _GATEKEEPER_LLS_CACHE_H_
+
+#include "gatekeeper_lls.h"
+#include "gatekeeper_net.h"
+
+/* Set to 1 to enable dumping the cache after it changes. */
+#define DEBUG_LLS_CACHE		0
+
+struct lls_cache_entry {
+	/* Ethernet address of this entry. */
+	struct ether_addr ha;
+
+	/*
+	 * Time that entry was added. When the entry is updated,
+	 * the timestamp is also updated.
+	 */
+	time_t            ts;
+
+	/* Port on which this entry exists. */
+	uint8_t           port_id;
+
+	/* Whether this entry has been marked as stale. */
+	int               stale;
+
+	/*
+	 * Whether this entry has a valid resolution. This is false
+	 * under two conditions:
+	 *   (1) If a block has expressed interest in an entry but
+	 *       it has not yet been resolved.
+	 *   (2) If a resolution for this entry becomes stale and
+	 *       requests are not answered but there is still
+	 *       interest in this entry.
+	 */
+	int               valid;
+
+	/* Number of requests to monitor this entry. */
+	uint32_t          refcnt;
+};
+
+int lls_cache_init(struct lls_config *lls_conf, struct lls_cache *cache);
+void lls_cache_destroy(struct lls_cache *cache);
+
+/* Process any updates that this cache has in its update ring. */
+unsigned int lls_cache_writer(struct lls_cache *cache);
+
+/*
+ * Scan the cache for stale entries and send requests or
+ * remove entries as needed.
+ */
+void lls_cache_scan(struct lls_cache *cache, struct gatekeeper_if *front,
+	struct gatekeeper_if *back);
+
+/*
+ * Look up the LLS cache entry for the key @ip_key, and if it exists
+ * fill @eth_addr with the corresponding entry.
+ *
+ * NOTE
+ *	This function assumes that is only asked to resolve keys
+ *	that have active interest. Otherwise, race conditions
+ *	could occur where an entry is looked up and quickly deleted.
+ */
+int lls_cache_get(struct lls_cache *cache, const void *ip_key,
+	struct ether_addr *eth_addr);
+
+/* Change the reference count of an entry according to @inc_refcnt. */
+int lls_cache_refcnt(struct lls_cache *cache, const void *ip_key,
+	uint8_t port_id, int inc_refcnt);
+
+/* Add an LLS entry for the key @ip_key and address @addr on @port_id. */
+int lls_cache_add(struct lls_cache *cache, const void *ip_key,
```

Comment:
This is the only comment that I didn't address in my rebase.

We could have the cache directly return the cache entries to the block so that the block can store them with the IP addresses its interested in. Even if the entry changes, the block will always have a direct reference to it and can use the updated Ethernet address. But this would only work for blocks that communicate with a constant number of IP addressess; otherwise, they would have to do a lookup of their saved entries anyway.

Another option might be to add a reference to the corresponding LLS cache entry for each entry in the LPM table, so that when a block does a lookup to find the next hop IP address, it also gets the associated LL address.

But other than those cases (a constant number of interests or doing two lookups in one), I'm not sure how an entry can always be there without needing a lookup.